### PR TITLE
#424 - Add Scala instrumentation and Scala API for Monix Tasks

### PR DESCRIPTION
**Pull request by GDownes**
### Overview

- Adds new Scala-based instrumentation project for Monix monix-2.13_3.4.0, which aims to instrument the Monix Scheduler trait, converting Runnables on execution to New Relic Token Aware Runnables.
- Adds new Scala-based API project newrelic-scala-monix-api for Monix newrelic-scala-monix-api, which aims to provide convenience for starting New Relic transactions and segments for Monix Tasks
- Adds new Scala-based instrumentation project for newrelic-scala-monix-api for Monix newrelic-scala-monix-api, which aims to instrument the New Relic Monix API to enable convenience method for starting New Relic transactions

### Related Github Issue
Include a link to the related GitHub issue, if applicable:
Scala instrumentation - Monix async framework support
https://github.com/newrelic/newrelic-java-agent/issues/424

### Testing
Added functional tests to newrelic-scala-monix-api project to test Monix task scenarios. Add functional test to Scala functional test workflow. Ran tests successfully locally and on GHA. Tested cross building to Scala 2.11, 2.12 and 2.13. Used gradle verify instrumentation task to verify instrumentation from Monix version 3.0.0 to lastest version 3.4.0 for Scala 2.11, 2.12, 2.13 and 3.

### Checks

[X] Are your contributions backwards compatible with relevant frameworks and APIs? Yes
[X] Does your code contain any breaking changes? Please describe. No
[X] Does your code introduce any new dependencies? Please describe. Yes new projects with related Monix dependencies

### CODE REVIEW

File: .github/workflows/GHA-Scala-Functional-Tests.yaml [hunk 1]
```diff
@@ -95,7 +95,7 @@ jobs:
           TEST_TYPE: functional
         run: |
           # Removed ":newrelic-cats-effect3-api:test" temporarily
-          ./gradlew --console=plain --parallel :newrelic-scala-api:test :newrelic-scala-cats-api:test :newrelic-scala-zio-api:test -Ptest${{ matrix.java-version }} --continue --scan -Dscan.uploadInBackground=false --build-cache
```
Comment (richard-gibson) [inline .github/workflows/GHA-Scala-Functional-Tests.yaml:98]: Chatted to Gerard about this, currently failing as the new submodule newrelic-scala-monix-api can't be found although tests are passing when run locally

File: instrumentation/newrelic-scala-monix-api/src/main/scala/com/newrelic/monix/api/Util.scala [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+      tracer => if(tracer == null) {
+        body
+      } else {
+        for {
+          _ <- Task.delay(NewRelic.getAgent.getTransaction)
+          res <- body.onErrorHandleWith(throwable => {
+            tracer.finish(throwable)
+            Task.raiseError(throwable)
+          })
+          _ <- Task.delay(tracer.finish(172, null))
```
Comment (richard-gibson) [inline instrumentation/newrelic-scala-monix-api/src/main/scala/com/newrelic/monix/api/Util.scala:22]: Want to change 172 to const at top of file just to make clear what the opcode is? I think this was done for CE3

File: newrelic-scala-monix-api/src/main/scala/com/newrelic/monix/api/TraceOps.scala [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    val segment = txn.startSegment(segmentName)
+    try {
+      block
+    } finally {
+      segment.end()
+    }
+  }
+
+  /**
+    * Creates a segment to capture metrics for a value in a context F[A] where there is a
```
Comment (richard-gibson) [inline newrelic-scala-monix-api/src/main/scala/com/newrelic/monix/api/TraceOps.scala:41]: Can you change `F[A]` to `Task[A]`

File: newrelic-scala-monix-api/src/main/scala/com/newrelic/monix/api/TraceOps.scala [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+      try {
+        f(t)
+      } finally {
+        segment.end()
+      }
+  }
+
+  /**
+    * Creates a segment to capture metrics for given asynchronous function of return type :F[S],
+    * When run the returned F[S] calls {@link com.newrelic.api.agent.Transaction# startSegment ( String )},
```
Comment (richard-gibson) [inline newrelic-scala-monix-api/src/main/scala/com/newrelic/monix/api/TraceOps.scala:102]: more comments to update

File: newrelic-scala-monix-api/src/main/scala/com/newrelic/monix/api/TraceOps.scala [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+  /**
+    * Wraps a given block of code so that a {@link com.newrelic.api.agent.Transaction} will be started and completed
+    * before and after the code is run.
+    * When this method is invoked within the context of an existing transaction this has no effect.
+    * The newly created {@link com.newrelic.api.agent.Transaction} will complete once the code block has been executed
+    * e.g. the code below will create a Transaction and with 2 segments <i>trace map IO</i> and <i>trace filter IO</i>
+    * <pre>
+    * txn {
+    * IO(1).map(traceFun("trace map IO")(i => i + 1))
```
Comment (richard-gibson) [inline newrelic-scala-monix-api/src/main/scala/com/newrelic/monix/api/TraceOps.scala:136]: IO => Task

File: .github/workflows/GHA-Scala-Functional-Tests.yaml (hunk 1 above)
Comment (GDownes) [reply .github/workflows/GHA-Scala-Functional-Tests.yaml:98]: GHA issues mentioned to @twcrone

**Review by richard-gibson — APPROVED** [review]

**Review by twcrone — APPROVED** [review]

**Pull request merged**